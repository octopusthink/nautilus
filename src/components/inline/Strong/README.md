`<Strong>` is used to mark a short span of text within body copy as important.

```jsx
import { Paragraph } from '@octopusthink/nautilus';

<Paragraph>
  <Strong>Rick James</Strong> was a great musician. He played the bass.
</Paragraph>
```

## Usage

Use a `<Strong>` component where you would usually use a `<strong>` tag in HTML.

### Use this component for...

- Highlighting important information.
- Helping users who are skimming to pick out key phrases and concepts.
- A piece of text requiring "global" emphasis (emphasis when scanning the page as a whole).

### Don't use this component for...

- A heading or subheading. Use a `<Heading>` or `<Subtitle>` instead.
- Highlighting a navigational link. Use a `<Link>` instead.
- Indicating a stress or shifted intonation in a passage of text. Use an `<Emphasis>` instead.
- A piece of text requiring "local" emphasis (emphasis when reading the passage). Use an `<Emphasis>` instead.

## Appearance

`<Strong>` spans are styled consistently throughout and have more visual weight than `<Emphasis>` spans. Colours are adjustable in your theme.

```jsx
import { Paragraph, Emphasis } from '@octopusthink/nautilus';

<Paragraph>
  The puppies <Strong>chased their tails</Strong> <Emphasis>almost</Emphasis> every single afternoon.
</Paragraph>
```

## Accessibility

Use care when highlighting text. Only highlight a few words at a time, and restrict it to key passages or important concepts.

Using `<Strong>` judiciously can help make your text more accessible. Consider using this component to highlight key passages in your text, especially when text is long. This can help users more quickly parse a text and pull out the important details.
