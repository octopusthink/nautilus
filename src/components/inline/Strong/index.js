import { css } from '@emotion/core';
import PropTypes from 'prop-types';
import React from 'react';

import { useTheme } from 'themes';

export const Strong = (props) => {
  const { children, ...otherProps } = props;

  const theme = useTheme();

  return (
    <strong
      css={css`
        background: ${theme.colors.text.strongBackground};
        color: ${theme.colors.text.strong};
        font-weight: ${theme.typography.fontWeights.bodyBold};
      `}
      {...otherProps}
    >
      {children}
    </strong>
  );
};

Strong.defaultProps = {
  children: undefined,
};

Strong.propTypes = {
  /** @ignore */
  children: PropTypes.node,
};

export const { defaultProps, propTypes } = Strong;

export default Strong;
