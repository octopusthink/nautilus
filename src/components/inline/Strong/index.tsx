import { css } from '@emotion/react';
import React, { FunctionComponent } from 'react';

import { NautilusComponentProps } from '../../types';
import { useTheme } from '../../../themes';

interface Props extends NautilusComponentProps {}

const Strong: FunctionComponent<Props> = (props: Props) => {
  const { children, unstyled, ...otherProps } = props;

  const theme = useTheme();

  return (
    <strong
      css={
        unstyled
          ? undefined
          : css`
              background: ${theme.colors.text.strongBackground};
              color: ${theme.colors.text.strong};
              font-weight: ${theme.typography.fontWeights.bodyBold};
            `
      }
      {...otherProps}
    >
      {children}
    </strong>
  );
};

export default Strong;
