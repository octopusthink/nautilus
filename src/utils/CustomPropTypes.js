import { Children } from 'react';

const allowedChildren = (...types) => {
  return (propValue, key, componentName) => {
    let error = null;

    Children.forEach(propValue[key], (child) => {
      if (!types.includes(child.type)) {
        error = new Error(
          `${componentName} component only accepts the following components as children: ${types
            .map((element) => element.displayName || element)
            .join(', ')}.`,
        );
      }
    });

    return error;
  };
};

export default {
  allowedChildren,
};
