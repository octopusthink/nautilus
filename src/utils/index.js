export { default as CustomPropTypes } from './CustomPropTypes';
