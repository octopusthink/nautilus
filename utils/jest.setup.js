// import axios from 'axios';
// import AxiosMockAdapter from 'axios-mock-adapter';
// import { JSDOM } from 'jsdom';
// eslint-disable-next-line import/no-extraneous-dependencies
import 'jest-axe/extend-expect';
// eslint-disable-next-line import/no-extraneous-dependencies
import '@testing-library/jest-dom/extend-expect';

global.window.matchMedia = jest.fn().mockImplementation((query) => ({
  matches: false,
  media: query,
  onchange: null,
  addListener: jest.fn(),
  removeListener: jest.fn(),
}));

jest.spyOn(global.console, 'error');
jest.spyOn(global.console, 'warn');
jest.spyOn(global.console, 'log');
jest.spyOn(global.console, 'info');
jest.spyOn(global.console, 'debug');

beforeEach(() => {
  global.console.error.mockClear();
  global.console.warn.mockClear();
  global.console.log.mockClear();
  global.console.info.mockClear();
  global.console.debug.mockClear();
});
