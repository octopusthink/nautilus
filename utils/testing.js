// We only use react-testing-library in development, so ignore this linting
// error.
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createMemoryHistory } from 'history';
import PropTypes from 'prop-types';
import React from 'react';
import { Router } from 'react-router-dom';

import { Nautilus } from 'components';
import { nautilus as nautilusDefaultTheme, themePropTypes } from 'themes';

// Wrap `react-testing-library`'s `render` function with our providers.
const NautilusProviders = ({ children, theme }) => {
  return <Nautilus theme={theme}>{children}</Nautilus>;
};

NautilusProviders.defaultProps = {
  children: undefined,
  theme: nautilusDefaultTheme,
};

NautilusProviders.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** @ignore */
  theme: themePropTypes,
};

const customRender = (ui, options) => {
  return render(ui, { wrapper: NautilusProviders, ...options });
};

// this is a handy function that I would utilize for any component
// that relies on the router being in context
export const renderWithRouter = (
  ui,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  } = {},
) => {
  return {
    ...customRender(<Router history={history}>{ui}</Router>),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
  };
};

export const muteConsole = ({ times, type } = {}) => {
  Array.from({ length: times }).forEach(() => {
    global.console[type].mockImplementationOnce(() => jest.fn());
  });
};

// Export jest-axe helpers.
export { axe } from 'jest-axe';

// Re-export everything else as normal.
export * from '@testing-library/react';

// override render method
export { customRender as render };
