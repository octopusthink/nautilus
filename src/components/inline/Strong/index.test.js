import React from 'react';

import { render } from 'utils/testing';

import Strong from '.';

describe('Strong', () => {
  it('should render a <strong> tag', () => {
    const { container } = render(<Strong>Hello</Strong>);

    expect(container.firstChild.tagName).toEqual('STRONG');
  });

  it('should output its children', () => {
    const { getByTestId } = render(
      <Strong>
        <span data-testid="child" />
      </Strong>,
    );

    expect(getByTestId('child')).toBeDefined();
  });

  it('should accept and pass through other props', () => {
    const { getByTestId } = render(
      <Strong className="custom-class" data-testid="myText">
        hello
      </Strong>,
    );

    expect(getByTestId('myText').classList).toContain('custom-class');
  });
});
