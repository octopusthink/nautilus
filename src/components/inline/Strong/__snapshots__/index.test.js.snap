// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Strong should match styles 1`] = `
.emotion-0 {
  background: #faebbb;
  color: #181b1c;
  font-weight: 600;
}

<div>
  <strong
    class="emotion-0 emotion-1"
  >
    hello
  </strong>
</div>
`;
